Description: Address FTBFS with gcc-8 due to format-overflow
Author: tony mancill <tmancill@debian.org>
--- zookeeper-client/zookeeper-client-c/src/zookeeper.c	2019-09-26 17:32:56.930700994 +0200
+++ zookeeper-client/zookeeper-client-c/src/zookeeper.c.new	2019-09-26 17:32:46.970795829 +0200
@@ -4381,7 +4381,7 @@
 
 static const char* format_endpoint_info(const struct sockaddr_storage* ep)
 {
-    static char buf[128] = { 0 };
+    static char buf[128 + 6];	// include space for the port :xxxxx
     char addrstr[128] = { 0 };
     void *inaddr;
 #ifdef _WIN32

