From 2d056b8da3fb3ee4480b7849eefae6637e9ed22c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ha=C3=AFkel=20Gu=C3=A9mar?= <hguemar@fedoraproject.org>
Date: Mon, 8 Dec 2014 10:04:18 +0100
Subject: [PATCH 4/5] redis 2.8.18: disable test failed on slow machine

---
 tests/test_helper.tcl | 1 -
 1 file changed, 1 deletion(-)

diff --git a/tests/test_helper.tcl b/tests/test_helper.tcl
index d1ebde1..369ebc2 100644
--- a/tests/test_helper.tcl
+++ b/tests/test_helper.tcl
@@ -33,7 +33,6 @@ set ::all_tests {
     integration/replication-2
     integration/replication-3
     integration/replication-4
-    integration/replication-psync
     integration/aof
     integration/rdb
     integration/convert-zipmap-hash-on-load
-- 
2.1.0

